source/manage/logging.rst

```diff
@@ -0,0 +1,481 @@
+    - Existing tools which ingest or parse audit log records may need to be modified.
+
+----
+
+Advanced logging
+-----------------
+
+.. include:: ../_static/badges/ent-only.rst
+  :start-after: :nosearch:
+
+System admins can output log records to any combination of `console <#console-target-configuration-options>`__, `local file <#file-target-configuration-options>`__, `syslog <#syslog-target-configuration-options>`__, and `TCP socket <#tcp-target-configuration-options>`__ targets. Each output target features additional configuration options you can customize for your Mattermost deployment.
+
+.. tip::
+    
+    - From Mattermost v9.3, system admins can configure advanced logging options in the System Console using multi-line JSON by going to **Environment > Logging**.
+    - Alternatively, admins can configure advanced logging within the ``AdvancedLoggingJSON`` section of the ``config.json`` file using multi-line JSON or escaped JSON as a string.
+    - Mattermost Team Edition customers can output audit log records to the console or a file.
+
+Advanced logging options can be configured to:
+
+- Enable trace logging for AD/LDAP to troubleshoot authentication issues.
+- Capture errors and panics in a separate, monitored file that triggers alerts.
+- Capture production debug and error logs in a separate file with log file rotation to reproduce issues.
+- Audit every API endpoint accessed during a user workflow.
+
+Configuring advanced logging includes the following steps:
+
+- `Define audit log output <#define-audit-log-output>`__ as multi-line JSON or a filespec to another configuration file.
+- `Specify destination targets <#specify-destination-targets>`__, including any combination of `console <#console-target-configuration-options>`__, `local file <#file-target-configuration-options>`__, `syslog <#syslog-target-configuration-options>`__, and `TCP socket <#tcp-target-configuration-options>`__ targets.
+- `Configure format preferences <#specify-destination-targets>`__ for `plain <#plain-log-format-configuration-options>`__, `JSON <#json-log-format-configuration-options>`__, or `GELF <#gelf-log-format-configuration-options>`__ output.
+- `Configure log levels & events <#configure-log-levels-and-events>`__
+- `Configure target-specific settings <#configure-target-specific-settings>`__
+
+Define audit log output
+~~~~~~~~~~~~~~~~~~~~~~~
+
+.. tabs::
+
+    .. tab:: Multi-line JSON
+
+        In the example below, file output is written to ``./logs/audit.log`` in plain text and includes all audit log levels & events. Older logs are kept for 1 day, and up to a total of 10 backup log files are kept at a time. Logs are rotated using gzip when the maximum size of the log file reaches 500 MB. A maximum of 1000 audit records can be queued/buffered while writing to the file.
+
+        .. code-block:: JSON
+
+            "AdvancedLoggingJSON": {
+                "file_1": {
```

Comment:
I get an error using this right now, trying to debug.

```
2023-12-13 15:21:22 {"timestamp":"2023-12-13 23:21:22.262 Z","level":"debug","msg":"Loaded audit configuration","caller":"app/audit.go:122","source":"{\"file_1\":{\"Type\":\"file\",\"Format\":\"plain\",\"Levels\":[{\"id\":100,\"name\":\"audit-api\"},{\"id\":101,\"name\":\"audit-content\"},{\"id\":102,\"name\":\"audit-permissions\"},{\"id\":103,\"name\":\"audit-cli\"}],\"Options\":{\"Compress\":true,\"Filename\":\"audit.log\",\"MaxAgeDays\":1,\"MaxBackups\":10,\"MaxSizeMB\":500},\"MaxQueueSize\":1000}}"}
2023-12-13 15:21:22 {"timestamp":"2023-12-13 23:21:22.262 Z","level":"error","msg":"Error configuring audit","caller":"app/server.go:455","error":"error creating log target _defAudit: invalid file target options: filename cannot be empty"}
2023-12-13 15:21:22 {"timestamp":"2023-12-13 23:21:22.262 Z","level":"debug","msg":"Loaded audit configuration","source":"{\"file_1\":{\"Type\":\"file\",\"Format\":\"plain\",\"Levels\":[{\"id\":100,\"name\":\"audit-api\"},{\"id\":101,\"name\":\"audit-content\"},{\"id\":102,\"name\":\"audit-permissions\"},{\"id\":103,\"name\":\"audit-cli\"}],\"Options\":{\"Compress\":true,\"Filename\":\"audit.log\",\"MaxAgeDays\":1,\"MaxBackups\":10,\"MaxSizeMB\":500},\"MaxQueueSize\":1000}}"}
2023-12-13 15:21:22 {"timestamp":"2023-12-13 23:21:22.262 Z","level":"error","msg":"Error configuring audit","error":"error creating log target _defAudit: invalid file target options: filename cannot be empty","stacktrace":[{"Function":"github.com/mattermost/mattermost/server/v8/channels/app.NewServer","File":"github.com/mattermost/mattermost/server/v8/channels/app/server.go","Line":455},{"Function":"github.com/mattermost/mattermost/server/v8/cmd/mattermost/commands.runServer","File":"github.com/mattermost/mattermost/server/v8/cmd/mattermost/commands/server.go","Line":74},{"Function":"github.com/mattermost/mattermost/server/v8/cmd/mattermost/commands.serverCmdF","File":"github.com/mattermost/mattermost/server/v8/cmd/mattermost/commands/server.go","Line":58},{"Function":"github.com/spf13/cobra.(*Command).execute","File":"github.com/spf13/cobra@v1.7.0/command.go","Line":940},{"Function":"github.com/spf13/cobra.(*Command).ExecuteC","File":"github.com/spf13/cobra@v1.7.0/command.go","Line":1068},{"Function":"github.com/spf13/cobra.(*Command).Execute","File":"github.com/spf13/cobra@v1.7.0/command.go","Line":992},{"Function":"github.com/mattermost/mattermost/server/v8/cmd/mattermost/commands.Run","File":"github.com/mattermost/mattermost/server/v8/cmd/mattermost/commands/root.go","Line":14},{"Function":"main.main","File":"github.com/mattermost/mattermost/server/v8/cmd/mattermost/main.go","Line":20},{"Function":"runtime.main","File":"runtime/proc.go","Line":250},{"Function":"runtime.goexit","File":"runtime/asm_amd64.s","Line":1598}]}
```

Reply:
`ExperimentalAuditSettings.FileName` must be defined for it to read this, even though it's being set in the advanced logging config.

@wiggin77 - Is this a bug or expected?

Reply:
I can break it with the below:

```json
"ExperimentalAuditSettings": {
        "FileEnabled": true,
        "FileName": "",
        "FileMaxSizeMB": 100,
        "FileMaxAgeDays": 0,
        "FileMaxBackups": 0,
        "FileCompress": false,
        "FileMaxQueueSize": 1000,
        "AdvancedLoggingJSON": {
            "file_1": {
                "Type": "file",
                "Format": "plain",
                "Levels": [
                    {
                        "id": 100,
                        "name": "audit-api"
                    },
                    {
                        "id": 101,
                        "name": "audit-content"
                    },
                    {
                        "id": 102,
                        "name": "audit-permissions"
                    },
                    {
                        "id": 103,
                        "name": "audit-cli"
                    }
                ],
                "Options": {
                    "Compress": true,
                    "Filename": "audit.log",
                    "MaxAgeDays": 1,
                    "MaxBackups": 10,
                    "MaxSizeMB": 500
                },
                "MaxQueueSize": 1000
            }
        },
        "AdvancedLoggingConfig": ""
    },
```

Reply:
```
"FileEnabled": true,
"FileName": "",
```
@coltoneshaw   Default file logging have been enabled but no filename provided.

Reply:
Set `"FileEnabled": false,` and that config should work.

Reply:
Error was resolved offline.  @coltoneshaw recommended that samples use `./logs/` as the path.